### PR TITLE
Fixed bug in hash method for factor class

### PR DESCRIPTION
**Pull request by khalibartan**
The earlier method took order into account .see #574 
Ping @ankurankan , @yashu-seth I don't like this implementation much. Can you suggest any other ?

### CODE REVIEW

File: pgmpy/factors/Factor.py [hunk 1]
```diff
@@ -770,9 +770,14 @@ def __eq__(self, other):
                 return True
 
     def __hash__(self):
-        return hash(str(self.variables) + str(self.cardinality.tolist()) +
-                    str(self.values.astype('float').tolist()))
-
+        value_index = 0
```
Comment (ankurankan) [inline pgmpy/factors/Factor.py]: @khalibartan Check how we change the order of variables in other methods by swapping axes. And before computing the hash simply just sort the variables and then compute the hash.

Comment (khalibartan) [inline pgmpy/factors/Factor.py]: @ankurankan I was also thinking for the same , but I believe this method happens to be faster .

**Comment by khalibartan** [comment]
@ankurankan There is an issue with `__eq__` method. The problem is that you are swaping variables in inplace. So after applying `__eq__` method, representation changes. Here is the snippet for better understanding

```
In [1]: from pgmpy.factors import Factor

In [2]: phi = Factor([('x',1),('x',2),('x',3)],[2,2,2],range(8))

In [3]: phi1 = Factor(['x3','x1','x2'],[2,2,2],[0,2,4,6,1,3,5,7])

In [4]: phi = Factor(['x1','x2','x3'],[2,2,2],range(8))

In [5]: print(phi1)
╒══════╤══════╤══════╤═════════════════╕
│ x3   │ x1   │ x2   │   phi(x3,x1,x2) │
╞══════╪══════╪══════╪═════════════════╡
│ x3_0 │ x1_0 │ x2_0 │          0.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_0 │ x1_0 │ x2_1 │          2.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_0 │ x1_1 │ x2_0 │          4.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_0 │ x1_1 │ x2_1 │          6.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_1 │ x1_0 │ x2_0 │          1.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_1 │ x1_0 │ x2_1 │          3.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_1 │ x1_1 │ x2_0 │          5.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x3_1 │ x1_1 │ x2_1 │          7.0000 │
╘══════╧══════╧══════╧═════════════════╛

In [6]: phi == phi1
Out[6]: True

In [7]: print(phi1)
╒══════╤══════╤══════╤═════════════════╕
│ x1   │ x2   │ x3   │   phi(x1,x2,x3) │
╞══════╪══════╪══════╪═════════════════╡
│ x1_0 │ x2_0 │ x3_0 │          0.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_0 │ x2_0 │ x3_1 │          1.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_0 │ x2_1 │ x3_0 │          2.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_0 │ x2_1 │ x3_1 │          3.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_1 │ x2_0 │ x3_0 │          4.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_1 │ x2_0 │ x3_1 │          5.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_1 │ x2_1 │ x3_0 │          6.0000 │
├──────┼──────┼──────┼─────────────────┤
│ x1_1 │ x2_1 │ x3_1 │          7.0000 │
╘══════╧══════╧══════╧═════════════════╛
```

Look representation of phi1 has changed after applying equality .

**Comment by khalibartan** [comment]
@ankurankan Method has been improved, take a look at it and also the mention bug. Do we want this behaviour or shall I fix it.

**Comment by ankurankan** [comment]
@khalibartan Yes please fix that issue with `__eq__` method also. It might cause problem with the new `TabularCPD`.

**Pull request merged**